src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/AnsiStringMarshaller.cs

```diff
@@ -0,0 +1,89 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime.CompilerServices;
+
+namespace System.Runtime.InteropServices
+{
+    /// <summary>
+    /// Marshaller for ANSI strings
+    /// </summary>
+    [CLSCompliant(false)]
+    [CustomTypeMarshaller(typeof(string), BufferSize = 0x100,
+        Features = CustomTypeMarshallerFeatures.UnmanagedResources | CustomTypeMarshallerFeatures.TwoStageMarshalling | CustomTypeMarshallerFeatures.CallerAllocatedBuffer)]
+    public unsafe ref struct AnsiStringMarshaller
+    {
+        private byte* allocated;
+        private Span<byte> span;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AnsiStringMarshaller"/>.
+        /// </summary>
+        /// <param name="str">The string to marshal.</param>
+        public AnsiStringMarshaller(string str)
+            : this(str, default(Span<byte>))
+        { }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AnsiStringMarshaller"/>.
+        /// </summary>
+        /// <param name="str">The string to marshal.</param>
+        /// <param name="buffer">Buffer that may be used for marshalling.</param>
+        /// <remarks>
+        /// <seealso cref="CustomTypeMarshallerFeatures.CallerAllocatedBuffer"/>
+        /// </remarks>
+        public AnsiStringMarshaller(string str, Span<byte> buffer)
+        {
+            allocated = null;
+            span = default;
+            if (str is null)
+                return;
```

Comment:
Would it help at all with overheads to only do this default initialization in the case of str is null, and then below add allocated = null in the if block and span = default in the else block?  Or is the extra code required for that worse than the duplicative field writes?